File: src/main/java/xyz/phanta/cbtweaker/integration/mekanism/jei/JeiMekanismHeatIngredient.java
package xyz.phanta.cbtweaker.integration.mekanism.jei;

import io.github.phantamanta44.libnine.util.format.FormatUtils;
import mekanism.common.util.UnitDisplayUtils;
import net.minecraft.client.util.ITooltipFlag;
import xyz.phanta.cbtweaker.gui.CbtTextureResources;
import xyz.phanta.cbtweaker.integration.jei.ingredient.JeiIngredient;

import java.util.Collections;
import java.util.List;

public class JeiMekanismHeatIngredient implements JeiIngredient<Double> {

    private final double amount;
    private final boolean isRate;
    private final Role role;

    public JeiMekanismHeatIngredient(double amount, boolean isRate, Role role) {
        this.amount = amount;
        this.isRate = isRate;
        this.role = role;
    }

    public double getAmount() {
        return amount;
    }

    @Override
    public List<Double> getIngredients() {
        return Collections.singletonList(amount);
    }

    @Override
    public Role getRole() {
        return role;
    }

    @Override
    public void renderIcon(int x, int y, Double ingredient) {
        CbtTextureResources.MEKANISM_ICON_HEAT.draw(x, y);
    }

    @Override
    public void getTooltip(Double ingredient, List<String> tooltip, ITooltipFlag tooltipFlags) {
        tooltip.add(isRate ? FormatUtils.formatSI(amount, "K/t")
                : FormatUtils.formatSI(UnitDisplayUtils.TemperatureUnit.AMBIENT.convertToK(amount, true), "K"));
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/compat/mekanism/jei/JeiMekanismHeatIngredient.kt
package st.evening.mc.cbtweaker.compat.mekanism.jei

import net.minecraft.client.util.ITooltipFlag
import net.minecraft.util.text.TextFormatting
import st.evening.mc.cbtweaker.compat.jei.ingredient.JeiIngredient
import st.evening.mc.cbtweaker.compat.mekanism.MekanismCompat
import st.evening.mc.cbtweaker.gui.CbtGuiResources
import st.evening.mc.prelude.api.gui.drawable.drawFullSize
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.game.RequireMod
import st.evening.mc.prelude.api.util.text.toStringPercentage
import st.evening.mc.prelude.api.util.text.toStringSI

@RequireMod(MekanismCompat.MOD_ID)
class JeiMekanismHeatIngredient(
    val amount: Double,
    val unitName: String,
    override val role: JeiIngredient.Role,
    val chance: Float = 1F
) : JeiIngredient<Double> {
    constructor(amount: Double, isRate: Boolean, role: JeiIngredient.Role, chance: Float = 1F) :
        this(amount, if (isRate) "J/t" else "J", role, chance) // are these actually joules...?

    override fun getIngredients(): List<Double> = listOf(amount)

    @ClientSide.Physical
    override fun drawIcon(x: Int, y: Int, ingredient: Double, partialTicks: Float) {
        CbtGuiResources.MEKANISM_ICON_HEAT.drawFullSize(partialTicks, x, y)
    }

    @ClientSide.Physical
    override fun getTooltip(ingredient: Double, tooltip: MutableList<String>, tooltipFlags: ITooltipFlag) {
        tooltip += amount.toStringSI(unitName)
        if (chance < 1F) {
            tooltip += "${TextFormatting.GOLD}(${chance.toStringPercentage()})"
        }
    }
}
